File: codegen/cpp/hyperbolic.cpp
// 3D Projective Geometric Algebra
// Written by a generator written by enki.
#include <stdio.h>
#include <cmath>
#include <array>

#define PI 3.14159265358979323846

static const char* basis[] = { "1","e1" };

class HYPERBOLIC {
  public:
    HYPERBOLIC ()  { std::fill( mvec, mvec + sizeof( mvec )/4, 0.0f ); }
    HYPERBOLIC (float f, int idx=0) { std::fill( mvec, mvec + sizeof( mvec )/4, 0.0f ); mvec[idx] = f; }
    float& operator [] (size_t idx) { return mvec[idx]; }
    const float& operator [] (size_t idx) const { return mvec[idx]; }
    HYPERBOLIC log () { int n=0; for (int i=0,j=0;i<2;i++) if (mvec[i]!=0.0f) { n++; printf("%s%0.7g%s",(j>0)?" + ":"",mvec[i],(i==0)?"":basis[i]); j++; };if (n==0) printf("0");  printf("\n"); return *this; }
    HYPERBOLIC Conjugate(); 
    HYPERBOLIC Involute();
    float norm();
    float inorm();
    HYPERBOLIC normalized();
  private:  
    float mvec[2];
};


//***********************
// HYPERBOLIC.Reverse : res = ~a
// Reverse the order of the basis blades.
//***********************
inline HYPERBOLIC operator ~ (const HYPERBOLIC &a) {
  HYPERBOLIC res;
  res[0]=a[0];
  res[1]=a[1];
  return res;
};

//***********************
// HYPERBOLIC.Dual : res = !a
// Poincare duality operator.
//***********************
inline HYPERBOLIC operator ! (const HYPERBOLIC &a) {
  HYPERBOLIC res;
  res[0]=a[1];
  res[1]=a[0];
  return res;
};

//***********************
// HYPERBOLIC.Conjugate : res = a.Conjugate()
// Clifford Conjugation
//***********************
inline HYPERBOLIC HYPERBOLIC::Conjugate () {
  HYPERBOLIC res;
  res[0]=this->mvec[0];
  res[1]=-this->mvec[1];
  return res;
};

//***********************
// HYPERBOLIC.Involute : res = a.Involute()
// Main involution
//***********************
inline HYPERBOLIC HYPERBOLIC::Involute () {
  HYPERBOLIC res;
  res[0]=this->mvec[0];
  res[1]=-this->mvec[1];
  return res;
};

//***********************
// HYPERBOLIC.Mul : res = a * b 
// The geometric product.
//***********************
inline HYPERBOLIC operator * (const HYPERBOLIC &a, const HYPERBOLIC &b) {
  HYPERBOLIC res;
  res[0]=b[0]*a[0]+b[1]*a[1];
  res[1]=b[1]*a[0]+b[0]*a[1];
  return res;
};

//***********************
// HYPERBOLIC.Wedge : res = a ^ b 
// The outer product. (MEET)
//***********************
inline HYPERBOLIC operator ^ (const HYPERBOLIC &a, const HYPERBOLIC &b) {
  HYPERBOLIC res;
  res[0]=b[0]*a[0];
  res[1]=b[1]*a[0]+b[0]*a[1];
  return res;
};

//***********************
// HYPERBOLIC.Vee : res = a & b 
// The regressive product. (JOIN)
//***********************
inline HYPERBOLIC operator & (const HYPERBOLIC &a, const HYPERBOLIC &b) {
  HYPERBOLIC res;
  res[1]=b[1]*a[1];
  res[0]=b[0]*a[1]+b[1]*a[0];
  return res;
};

//***********************
// HYPERBOLIC.Dot : res = a | b 
// The inner product.
//***********************
inline HYPERBOLIC operator | (const HYPERBOLIC &a, const HYPERBOLIC &b) {
  HYPERBOLIC res;
  res[0]=b[0]*a[0]+b[1]*a[1];
  res[1]=b[1]*a[0]+b[0]*a[1];
  return res;
};

//***********************
// HYPERBOLIC.Add : res = a + b 
// Multivector addition
//***********************
inline HYPERBOLIC operator + (const HYPERBOLIC &a, const HYPERBOLIC &b) {
  HYPERBOLIC res;
      res[0] = a[0]+b[0];
    res[1] = a[1]+b[1];
  return res;
};

//***********************
// HYPERBOLIC.Sub : res = a - b 
// Multivector subtraction
//***********************
inline HYPERBOLIC operator - (const HYPERBOLIC &a, const HYPERBOLIC &b) {
  HYPERBOLIC res;
      res[0] = a[0]-b[0];
    res[1] = a[1]-b[1];
  return res;
};

//***********************
// HYPERBOLIC.smul : res = a * b 
// scalar/multivector multiplication
//***********************
inline HYPERBOLIC operator * (const float &a, const HYPERBOLIC &b) {
  HYPERBOLIC res;
      res[0] = a*b[0];
    res[1] = a*b[1];
  return res;
};

//***********************
// HYPERBOLIC.muls : res = a * b 
// multivector/scalar multiplication
//***********************
inline HYPERBOLIC operator * (const HYPERBOLIC &a, const float &b) {
  HYPERBOLIC res;
      res[0] = a[0]*b;
    res[1] = a[1]*b;
  return res;
};

//***********************
// HYPERBOLIC.sadd : res = a + b 
// scalar/multivector addition
//***********************
inline HYPERBOLIC operator + (const float &a, const HYPERBOLIC &b) {
  HYPERBOLIC res;
    res[0] = a+b[0];
      res[1] = b[1];
  return res;
};

//***********************
// HYPERBOLIC.adds : res = a + b 
// multivector/scalar addition
//***********************
inline HYPERBOLIC operator + (const HYPERBOLIC &a, const float &b) {
  HYPERBOLIC res;
    res[0] = a[0]+b;
      res[1] = a[1];
  return res;
};


inline float HYPERBOLIC::norm() { return sqrt(abs(((*this)*Conjugate()).mvec[0])); }
inline float HYPERBOLIC::inorm() { return (!(*this)).norm(); }
inline HYPERBOLIC HYPERBOLIC::normalized() { return (*this) * (1/norm()); }


static HYPERBOLIC e1(1.0f,1);


int main (int argc, char **argv) {
  
  printf("e1*e1         : "); (e1*e1).log();
  printf("pss           : "); e1.log();
  printf("pss*pss       : "); (e1*e1).log();

  return 0;
}

File: codegen/rust/hyperbolic.rs
// 3D Projective Geometric Algebra
// Written by a generator written by enki.
#![allow(unused_imports)]
#![allow(dead_code)]
#![allow(non_upper_case_globals)]
#![allow(non_snake_case)]
#![allow(non_camel_case_types)]
#![feature(const_slice_len)]

use std::fmt;
use std::ops::{Index,IndexMut,Add,Sub,Mul,BitAnd,BitOr,BitXor,Not};

type float_t = f64;

// use std::f64::consts::PI;
const PI: float_t = 3.14159265358979323846;

const basis: &'static [&'static str] = &[ "1","e1" ];
const basis_count: usize = basis.len();

#[derive(Default,Debug,Clone,Copy,PartialEq)]
struct HYPERBOLIC {
    mvec: [float_t; basis_count]
}

impl HYPERBOLIC {
    pub const fn zero() -> Self {
        Self {
            mvec: [0.0; basis_count]
        }
    }

    pub const fn new(f: float_t, idx: usize) -> Self {
        let mut ret = Self::zero();
        ret.mvec[idx] = f;
        ret
    }
}

// basis vectors are available as global constants.
const e1: HYPERBOLIC = HYPERBOLIC::new(1.0, 1);

impl Index<usize> for HYPERBOLIC {
    type Output = float_t;

    fn index<'a>(&'a self, index: usize) -> &'a Self::Output {
        &self.mvec[index]
    }
}

impl IndexMut<usize> for HYPERBOLIC {
    fn index_mut<'a>(&'a mut self, index: usize) -> &'a mut Self::Output {
        &mut self.mvec[index]
    }
}

impl fmt::Display for HYPERBOLIC {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut n = 0;
        let ret = self.mvec.iter().enumerate().filter_map(|(i, &coeff)| {
            if coeff > 0.00001 || coeff < -0.00001 {
                n = 1;
                Some(format!("{}{}", 
                        format!("{:.*}", 7, coeff).trim_end_matches('0').trim_end_matches('.'),
                        if i > 0 { basis[i] } else { "" }
                    )
                )
            } else {
                None
            }
        }).collect::<Vec<String>>().join(" + ");
        if n==0 { write!(f,"0") } else { write!(f, "{}", ret) }
    }
}

// Reverse
// Reverse the order of the basis blades.
impl HYPERBOLIC {
    pub fn Reverse(self: Self) -> HYPERBOLIC {
        let mut res = HYPERBOLIC::zero();
        let a = self;
        res[0]=a[0];
        res[1]=a[1];
        res
    }
}

// Dual
// Poincare duality operator.
impl HYPERBOLIC {
    pub fn Dual(self: Self) -> HYPERBOLIC {
        let mut res = HYPERBOLIC::zero();
        let a = self;
        res[0]=a[1];
        res[1]=a[0];
        res
    }
}

impl Not for HYPERBOLIC {
    type Output = HYPERBOLIC;

    fn not(self: Self) -> HYPERBOLIC {
        let mut res = HYPERBOLIC::zero();
        let a = self;
        res[0]=a[1];
        res[1]=a[0];
        res
    }
}

// Conjugate
// Clifford Conjugation
impl HYPERBOLIC {
    pub fn Conjugate(self: Self) -> HYPERBOLIC {
        let mut res = HYPERBOLIC::zero();
        let a = self;
        res[0]=a[0];
        res[1]=-a[1];
        res
    }
}

// Involute
// Main involution
impl HYPERBOLIC {
    pub fn Involute(self: Self) -> HYPERBOLIC {
        let mut res = HYPERBOLIC::zero();
        let a = self;
        res[0]=a[0];
        res[1]=-a[1];
        res
    }
}

// Mul
// The geometric product.
impl Mul for HYPERBOLIC {
    type Output = HYPERBOLIC;

    fn mul(self: HYPERBOLIC, b: HYPERBOLIC) -> HYPERBOLIC {
        let mut res = HYPERBOLIC::zero();
        let a = self;
        res[0]=b[0]*a[0]+b[1]*a[1];
		res[1]=b[1]*a[0]+b[0]*a[1];
        res
    }
}

// Wedge
// The outer product. (MEET)
impl BitXor for HYPERBOLIC {
    type Output = HYPERBOLIC;

    fn bitxor(self: HYPERBOLIC, b: HYPERBOLIC) -> HYPERBOLIC {
        let mut res = HYPERBOLIC::zero();
        let a = self;
        res[0]=b[0]*a[0];
		res[1]=b[1]*a[0]+b[0]*a[1];
        res
    }
}

// Vee
// The regressive product. (JOIN)
impl BitAnd for HYPERBOLIC {
    type Output = HYPERBOLIC;

    fn bitand(self: HYPERBOLIC, b: HYPERBOLIC) -> HYPERBOLIC {
        let mut res = HYPERBOLIC::zero();
        let a = self;
        res[1]=b[1]*a[1];
		res[0]=b[0]*a[1]+b[1]*a[0];
        res
    }
}

// Dot
// The inner product.
impl BitOr for HYPERBOLIC {
    type Output = HYPERBOLIC;

    fn bitor(self: HYPERBOLIC, b: HYPERBOLIC) -> HYPERBOLIC {
        let mut res = HYPERBOLIC::zero();
        let a = self;
        res[0]=b[0]*a[0]+b[1]*a[1];
		res[1]=b[1]*a[0]+b[0]*a[1];
        res
    }
}

// Add
// Multivector addition
impl Add for HYPERBOLIC {
    type Output = HYPERBOLIC;

    fn add(self: HYPERBOLIC, b: HYPERBOLIC) -> HYPERBOLIC {
        let mut res = HYPERBOLIC::zero();
        let a = self;
        res[0] = a[0]+b[0];
		res[1] = a[1]+b[1];
        res
    }
}

// Sub
// Multivector subtraction
impl Sub for HYPERBOLIC {
    type Output = HYPERBOLIC;

    fn sub(self: HYPERBOLIC, b: HYPERBOLIC) -> HYPERBOLIC {
        let mut res = HYPERBOLIC::zero();
        let a = self;
        res[0] = a[0]-b[0];
		res[1] = a[1]-b[1];
        res
    }
}

// smul
// scalar/multivector multiplication
impl Mul<HYPERBOLIC> for float_t {
    type Output = HYPERBOLIC;

    fn mul(self: float_t, b: HYPERBOLIC) -> HYPERBOLIC {
        let mut res = HYPERBOLIC::zero();
        let a = self;
        res[0] = a*b[0];
        res[1] = a*b[1];
        res
    }
}

// muls
// multivector/scalar multiplication
impl Mul<float_t> for HYPERBOLIC {
    type Output = HYPERBOLIC;

    fn mul(self: HYPERBOLIC, b: float_t) -> HYPERBOLIC {
        let mut res = HYPERBOLIC::zero();
        let a = self;
        res[0] = a[0]*b;
        res[1] = a[1]*b;
        res
    }
    }

// sadd
// scalar/multivector addition
impl Add<HYPERBOLIC> for float_t {
    type Output = HYPERBOLIC;

    fn add(self: float_t, b: HYPERBOLIC) -> HYPERBOLIC {
        let mut res = HYPERBOLIC::zero();
        let a = self;
        res[0] = a+b[0];
        res[1] = b[1];
        res
    }
}

// adds
// multivector/scalar addition
impl Add<float_t> for HYPERBOLIC {
    type Output = HYPERBOLIC;

    fn add(self: HYPERBOLIC, b: float_t) -> HYPERBOLIC {
        let mut res = HYPERBOLIC::zero();
        let a = self;
        res[0] = a[0]+b;
        res[1] = a[1];
        res
    }
    }

impl HYPERBOLIC {
    pub fn norm(self: Self) -> float_t {
        let scalar_part = (self * self.Conjugate())[0];

        scalar_part.abs().sqrt()
    }

    pub fn inorm(self: Self) -> float_t {
        self.Dual().norm()
    }

    pub fn normalized(self: Self) -> Self {
        self * (1.0 / self.norm())
    }
    
    

}


fn main() {

  println!("e1*e1         : {}", e1 * e1);
  println!("pss           : {}", e1);
  println!("pss*pss       : {}", e1*e1);

}